From 97eaea01c61eb9fea578c844f9db387582557d90 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Thu, 17 Feb 2022 17:25:31 +0100
Subject: [PATCH 064/841] feat(st-regulator): introduce STM32 GPIO regulator

Introduce STM32 GPIO regulator, supporting only
enable/disable from a GPIO.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 0f602586887658483033cca80f5f26761b93b493 feat(st-regulator): introduce STM32 GPIO regulator]
Change-Id: I4489c698916a5dc632a9382e449e8d6a3f3f5ce3
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/240637
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/st/regulator/regulator_gpio.c | 166 ++++++++++++++++++++++++++
 include/drivers/st/regulator_gpio.h   |  12 ++
 2 files changed, 178 insertions(+)
 create mode 100644 drivers/st/regulator/regulator_gpio.c
 create mode 100644 include/drivers/st/regulator_gpio.h

diff --git a/drivers/st/regulator/regulator_gpio.c b/drivers/st/regulator/regulator_gpio.c
new file mode 100644
index 000000000..ebddb0f6e
--- /dev/null
+++ b/drivers/st/regulator/regulator_gpio.c
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2022, STMicroelectronics
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+
+#include <common/debug.h>
+#include <common/fdt_wrappers.h>
+#include <drivers/st/regulator.h>
+#include <drivers/st/regulator_gpio.h>
+#include <drivers/st/stm32_gpio.h>
+#include <libfdt.h>
+
+#ifndef PLAT_NB_GPIO_REGUS
+#error "Missing PLAT_NB_GPIO_REGUS"
+#endif
+
+#define MAX_NAME_LEN 32
+
+#define REGU_DISABLED 0
+#define REGU_ENABLED 1
+
+#define GPIO_DESC_SIZE 3U
+
+struct gpio_config {
+	uint32_t bank;
+	uint32_t pin;
+	uint32_t config;
+	uint8_t status;
+};
+
+struct gpio_regu_data {
+	char name[MAX_NAME_LEN];
+	struct regul_description desc;
+	struct gpio_config en_gpio;
+	bool active_high;
+};
+
+static struct gpio_regu_data data[PLAT_NB_GPIO_REGUS];
+
+static int gpio_set_state(const struct regul_description *desc, bool state)
+{
+	struct gpio_regu_data *d = (struct gpio_regu_data *)desc->driver_data;
+	enum gpio_level level = GPIO_LEVEL_LOW;
+
+	VERBOSE("regul %s set state=%d\n", d->name, state);
+
+	if ((d->active_high && state) || (!d->active_high && !state)) {
+		level = GPIO_LEVEL_HIGH;
+	}
+
+	set_gpio_level(d->en_gpio.bank, d->en_gpio.pin, level);
+
+	return 0;
+}
+
+static int gpio_get_state(const struct regul_description *desc)
+{
+	struct gpio_regu_data *d = (struct gpio_regu_data *)desc->driver_data;
+	enum gpio_level level = get_gpio_level(d->en_gpio.bank, d->en_gpio.pin);
+
+	if ((d->active_high && level == GPIO_LEVEL_HIGH) ||
+	    (!d->active_high && level == GPIO_LEVEL_LOW)) {
+		return REGU_ENABLED;
+	}
+
+	return REGU_DISABLED;
+}
+
+static struct regul_ops gpio_regu_ops = {
+	.set_state = gpio_set_state,
+	.get_state = gpio_get_state,
+};
+
+static int read_gpio_config_from_dt(const void *fdt, int node, char *name,
+				    struct gpio_config *gpio)
+{
+	uint32_t utable[GPIO_DESC_SIZE];
+	int ret;
+
+	gpio->status = fdt_get_status(node);
+	if (gpio->status == DT_DISABLED) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	ret = fdt_read_uint32_array(fdt, node, name, GPIO_DESC_SIZE, utable);
+	if (ret != 0) {
+		ERROR("Missing gpio description\n");
+		return ret;
+	}
+
+	gpio->bank = utable[0];
+	gpio->pin = utable[1];
+	gpio->config = utable[2];
+
+	return 0;
+}
+
+int gpio_regulator_register(void)
+{
+	uint32_t count = 0U;
+	void *fdt;
+	int node = 0;
+
+	VERBOSE("gpio_regu reg init\n");
+
+	if (fdt_get_address(&fdt) == 0) {
+		return -FDT_ERR_NOTFOUND;
+	}
+
+	do {
+		int len __unused;
+		int ret;
+		struct gpio_regu_data *d = &data[count];
+		const char *reg_name;
+
+		node = fdt_node_offset_by_compatible(fdt, node, "st,stm32-regulator-gpio");
+		if (node < 0) {
+			break;
+		}
+
+		reg_name = fdt_get_name(fdt, node, NULL);
+
+		VERBOSE("register gpio_regu reg %s\n", reg_name);
+
+		if (count == PLAT_NB_GPIO_REGUS) {
+			WARN("Not enough place for regulators, PLAT_NB_GPIO_REGUS should be increased.\n");
+			return -ENOMEM;
+		}
+
+		len = snprintf(d->name, MAX_NAME_LEN - 1, "%s", reg_name);
+		assert((len > 0) && (len < (MAX_NAME_LEN - 1)));
+
+		d->desc.node_name = d->name;
+		d->desc.driver_data = d;
+		d->desc.ops = &gpio_regu_ops;
+
+		ret = read_gpio_config_from_dt(fdt, node, "st,enable-gpios", &d->en_gpio);
+		if (ret != 0) {
+			ERROR("Incorrect gpio regul description\n");
+			return ret;
+		}
+
+		VERBOSE("%s: en_gpio bank=%u pin=%u cfg=%x\n", d->name,
+			d->en_gpio.bank, d->en_gpio.pin, d->en_gpio.config);
+
+		set_gpio_config(d->en_gpio.bank, d->en_gpio.pin, d->en_gpio.config,
+				d->en_gpio.status);
+
+		if (fdt_getprop(fdt, node, "enable-active-high", NULL))
+			d->active_high = true;
+
+		ret = regulator_register(&d->desc, node);
+		if (ret != 0) {
+			WARN("Failed to register %s\n", reg_name);
+			return ret;
+		}
+
+		count++;
+	} while (node > 0);
+
+	return 0;
+}
diff --git a/include/drivers/st/regulator_gpio.h b/include/drivers/st/regulator_gpio.h
new file mode 100644
index 000000000..3362f7e24
--- /dev/null
+++ b/include/drivers/st/regulator_gpio.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2022, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef REGULATOR_GPIO_H
+#define REGULATOR_GPIO_H
+
+int gpio_regulator_register(void);
+
+#endif /* REGULATOR_GPIO_H */
-- 
2.47.2

